crates/gitbutler-core/src/virtual_branches/controller.rs

```diff
@@ -43,29 +44,33 @@ impl Controller {
         run_hooks: bool,
     ) -> Result<git2::Oid> {
         let _permit = self.semaphore.acquire().await;
-        self.with_verify_branch(project, |project_repository| {
-            let snapshot_tree = project_repository.project().prepare_snapshot();
-            let result =
-                super::commit(project_repository, branch_id, message, ownership, run_hooks)
-                    .map_err(Into::into);
-            let _ = snapshot_tree.and_then(|snapshot_tree| {
-                project_repository.project().snapshot_commit_creation(
-                    snapshot_tree,
-                    result.as_ref().err(),
-                    message.to_owned(),
-                    None,
-                )
-            });
-            result
-        })
+        let project_repository = open_with_verify(project)?;
```

Comment:
I see now - doing preparation is best done without a closure as it has to be at the beginning of a method anyway. This is definitely easier to look at than the closure version, and ~40 lines removed overall definitely prove this right as well.